### ISSUE TITLE
Circular reference array doesn't work for multiple relationships

### ISSUE THREAD

**Issue by tomxucnxa** [hyperledger/composer#3934]
Circular reference does work fine only if the relationship is single; but it will throw 'Error: Generated invalid JSON' if there are multiple relationships are pushed into the Array.

## Context

There are 2 assets definition as below:
```
asset Host{
--> Client[] clients optional
}
asset Client{
--> Host host optional
}
```
Once a 'ConnectionTransaction' is submitted, it will set the 'host' property of the Client instance, and then add the Client instance into the 'clients' property of the Host instance, what is an array.
It works fine, at the first time, we connect 'host1' with 'client1'; but it get failed when we connect 'host1' with 'client2', and an ''Error: Generated invalid JSON'' is thrown. Details as below:
```
Error: Generated invalid JSON:
//... Only the most important part remains
"clients":[{"$class":
"org.acme.reftest.Client","id":"c1","host":"resource:org.acme.reftest.Host#h1"}"resource:org.acme.reftest.Client#c2"]
```
Obviously, there should be a 'comma' before the "resource..." string, this should be the root cause.
The transaction should be consistent for both single instance and multiple instances in array.

## Expected Behavior

The transaction should be consistent for both single instance and multiple instances in array, for the circular reference usage.

## Actual Behavior

It works only for single instance, but fails when adding multiple relationships into the array.

## Possible Fix

I went through the code quickly, and **GUESS** it might be caused by jsongenerator.js line 206-209.
https://github.com/hyperledger/composer/blob/master/packages/composer-common/lib/serializer/jsongenerator.js
Method 'parameters.writer.writeComma()' should be called in this sect, likes below:
```
if (parameters.seenResources.has(fqi)) {
    let relationshipText = this.getRelationshipText(relationshipDeclaration, item);

    // Fix trying:
    parameters.writer.writeComma();

    parameters.writer.writeStringValue(relationshipText);
}
```
But I didn't debug or rebuild this project to verify fixing, hope anyone can look into it for further information and fixing.

## Steps to Reproduce

1. Install and start testref-business-network.bna (file attached).
2. Add asset Client, Host
```
{
  "$class": "org.acme.reftest.Client",
  "id": "c1"
}
{
  "$class": "org.acme.reftest.Client",
  "id": "c2"
}
{
  "$class": "org.acme.reftest.Host",
  "id": "h1"
}
```
3. Submit ConnectTransaction, to connect h1-c1. It will get succeed.
```
{
  "$class": "org.acme.reftest.ConnectTransaction",
  "host": "resource:org.acme.reftest.Host#h1",
  "client": "resource:org.acme.reftest.Client#c1"
}
```
4. Submit one more ConnectTransaction to connect h1-c2. It will get failed.
```
{
  "$class": "org.acme.reftest.ConnectTransaction",
  "host": "resource:org.acme.reftest.Host#h1",
  "client": "resource:org.acme.reftest.Client#c2"
}
```
**I think the key error message is: (file attached). The 'comma' is missed when if (parameters.seenResources.has(fqi)) ... '**
```
\"clients\":[{\"$class\":
\"org.acme.reftest.Client\",\"id\":\"c1\",\"host\":\"resource:org.acme.reftest.Host#h1\"}\"resource:org.acme.reftest.Client#c2\"]
```

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [x] [GitHub Issues](https://github.com/hyperledger/composer/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

https://github.com/hyperledger/composer/issues/3408

## Context

It seems to be a same issue. But I'd provide more details and a fix trying here.

## Your Environment

* Version used: Composer 0.19.2
* Environment name and version (e.g. Chrome 39, node.js 5.4): node v8.11.1
* Operating System and version (desktop or mobile): Ubuntu 16.04.4 LTS
* Link to your project: 2 attached files.

## Files Attached
[testref-business-network.bna.zip](https://github.com/hyperledger/composer/files/1957626/testref-business-network.bna.zip)
[err_h1_c2.txt](https://github.com/hyperledger/composer/files/1957627/err_h1_c2.txt)

**Comment by mahoney1** [comment]
hello,

thanks for posting :-) I suspect your problem is how you form your transaction JSON. Multiple relationships can be pushed to an array.

As you know `--> Client[] clients optional` is an array of relationships (ie resource identifiers) to the Client type, stored in a field called 'clients'. I'll provide details on that below.  The format, similar to what you're trying to achieve - ie store a full 'client' in an array of clients - would be  `Client[] clients optional` only.  I answered this question on Stack Overflow yesterday (see near the bottom) for a comparison of both of these definitions and the correct JSON forms, including the transaction code to achieve this -> https://stackoverflow.com/questions/50325197/hyperledger-composer-error-8-resource-exhausted

So, with a model (as you actually have an id field anyway in your data) of : 

asset Host identified by id {
o String id
--> Client[] clients optional
}

asset Client identified by id {
o String id
--> Host host optional
}

{
  "$class": "org.acme.reftest.Client",
  "id": "c1"
}
{
  "$class": "org.acme.reftest.Client",
  "id": "c2"
}
{
  "$class": "org.acme.reftest.Host",
  "id": "h1"
}

your transaction to update the 'Host' clients optional array should yield:

{
  "$class": "org.acme.reftest.Host",
  "id": "h1",
  "clients": [
    "resource:org.acme.reftest.Client#c1",
    "resource:org.acme.reftest.Client#c2"
  ]
}

Hope this helps.

**Comment by tomxucnxa** [comment]
Thanks for comments.
But the problem is that: Composer doesn't handle the **'circular' references array** correctly, and a unexpected exception was thrown.
I think the exception of 'invalid json' should not be generated, and Composer should process that, since it surely allow developer to use circular references array.

**Comment by mahoney1** [comment]
OK, I see, my apologies. it may indeed be a serializer bug that you've spotted - thanks for the pointers and the business network which I didn't see first time around. Will get someone to investigate.

**Comment by skarlekar** [comment]
I am having the same issue. I had the circular dependency working fine in v0.19.0. Now it is broken with v0.19.6 and also v0.19.7. 
Is there an ETA on when this fix will be released?
Additionally, how I do install version 0.19.0. Any help will be greatly appreciated.

**Comment by tomxucnxa** [comment]
@skarlekar I am not very sure but I'd suggest trying 'npm install -g composer-cli@0.19.0' ?

**Comment by skarlekar** [comment]
@tomxucnxa Thanks for the suggestion. I tried that but ran into a wall on that one as well. Apparently the version 019.0 of composer libraries depends upon version 1.10.1 of grpc. 

I tried npm install --save grpc@1.10.1 and reinstalling the composer libraries v0.19.0, but still get the same error.

**Comment by mbwhite** [comment]
Trying out the suggested fix looks good to me; locally testing showed one failure that turned out to be the test that should have caught this not doing so. So fixed that as well.

Going to get this run over the full tests now to see what happens.

**Comment by mbwhite** [comment]
@tomxucnxa @skarlekar  The fix as suggested is the #4098 PR...

**Comment by skarlekar** [comment]
@mbwhite Matthew, Thank you very much for the fix. How do I go about installing this fix? The Composer releases at https://github.com/hyperledger/composer/releases still show version 0.19.7 as the latest.
